// scanex-auth-control/rollup.config.js
import babel from 'rollup-plugin-babel';
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import css from 'rollup-plugin-css-porter';
import svelte from 'rollup-plugin-svelte';
import pkg from './package.json';

export default [
    {
        input: 'example.js',
        output: { 
            file: pkg.browser,
            format: 'iife',
            name: 'ns',
            sourcemap: true,
            globals: {
                leaflet: 'L'
            }
        },
        plugins: [
            svelte(),
            resolve(),
            commonjs(),
            css({dest: 'public/main.css', minify: false}),
            babel(),
        ],
    },
    {
        input: 'index.js',
        external: ['leaflet', 'scanex-translations'],	
        output: { 
            file: pkg.main,
            format: 'cjs',
            sourcemap: true,
            globals: {
                leaflet: 'L'
            }
        },
        plugins: [
            svelte(),
            resolve(),
            commonjs(),            
            css({dest: 'dist/scanex-auth-control.css', minify: false}),
            babel(),
        ],
    }
];
