// scanex-auth-control/src/AuthControl.js
import AuthWidget from './AuthWidget.svelte';
import L from 'leaflet';

let authWidget;

let AuthControl = L.Control.extend({
    includes: L.Evented.prototype,
    onAdd: function(map) {
        let container = L.DomUtil.create('div');
        authWidget = new AuthWidget({target: container});
        authWidget.$on('login', ({detail}) => {
            this.fire('login', {sync: detail});
        });
        authWidget.$on('logout', () => {
            this.fire('logout');
        });
        authWidget.$on('denied', () => this.fire('denied'));
        return container;
    },

    onRemove: function(map) {
        if(authWidget) {
            authWidget.$destroy();
        }
    },

    login: function () {
        if(authWidget) {
            authWidget.login();
        }
    },

    logout: function () {
        if(authWidget) {
            authWidget.logout();
        }
    }
});

const authControl = function(opts) {
    return new AuthControl(opts);
}

export {AuthControl, authControl};
